packages/firestore/src/local/indexeddb_remote_document_cache.ts

```diff
@@ -281,14 +282,24 @@ export class IndexedDbRemoteDocumentCache implements RemoteDocumentCache {
           return;
         }
 
-        const maybeDoc = this.serializer.fromDbRemoteDocument(dbRemoteDoc);
-        if (!query.path.isPrefixOf(maybeDoc.key.path)) {
-          control.done();
-        } else if (maybeDoc instanceof Document && query.matches(maybeDoc)) {
-          results = results.insert(maybeDoc.key, maybeDoc);
+        if (!query.path.isPrefixOf(new ResourcePath(key))) {
+            control.done();
+            return;
         }
+
+        const decodeDoc = new Promise<void>((resolve, reject) => {
```

Comment:
This could be a tad simpler:

```
Promise.resolve().then(() => {
     const maybeDoc = this.serializer.fromDbRemoteDocument(dbRemoteDoc);
     if (maybeDoc instanceof Document && query.matches(maybeDoc)) {
            results = results.insert(maybeDoc.key, maybeDoc);
       }
});
```

Also, I am not 100% certain that the code in `new Promise() { .... }` is already async - but the code that runs after a resolve() definitely is. It might be worth checking what the behavior is for the former.

Reply:
Done.